### PR TITLE
clusterworkspaces: add an APIBinding initializer

### PR DESCRIPTION
**Pull request by stevekuznetsov**
Depends on #1330 
Part of #1149 
/cc @sttts @davidfestal

### CODE REVIEW

File: pkg/reconciler/tenancy/initialization/apis/initialization/v1alpha1/types.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+// applies to all workspaces of the type that shares a name with this
+// object.
+//
+// +crd
+// +genclient
+// +genclient:nonNamespaced
+// +kubebuilder:resource:scope=Cluster,categories=kcp
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:subresource:status
+type APISet struct {
```
Comment (stevekuznetsov) [inline pkg/reconciler/tenancy/initialization/apis/initialization/v1alpha1/types.go:37]: Bikeshed begins :)

File: pkg/reconciler/tenancy/initialization/apibinder/bootstrap.go [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+		Spec: apisv1alpha1.APIExportSpec{
+			LatestResourceSchemas: []string{resourceSchema.Name},
+		},
+	}, metav1.CreateOptions{})
+	if err != nil && !k8serrors.IsAlreadyExists(err) {
+		return fmt.Errorf("could not create APIExport: %w", err)
+	}
+	klog.Infof("Created APIExport %q|%q", installInto.String(), export.Name)
+
+	if err := wait.ExponentialBackoffWithContext(ctx, wait.Backoff{
```
Comment (sttts) [inline pkg/reconciler/tenancy/initialization/apibinder/bootstrap.go:101]: Is this part of a controller reconcile? If yes, this is not a good idea as it blocks the controller. With only limited number of workers this would break the initializer controller quickly. In other words, do the backoff outside of the reconcile through the queue.

File: hack/update-codegen-clients.sh [hunk 3]
```diff
@@ -29,21 +29,32 @@ bash "${CODEGEN_PKG}"/generate-groups.sh "deepcopy,client,informer,lister" \
...
+  --trim-path-prefix github.com/kcp-dev/kcp &
+
+bash "${CODEGEN_PKG}"/generate-groups.sh "deepcopy,client,informer,lister" \
+  github.com/kcp-dev/kcp/pkg/reconciler/tenancy/initialization/client github.com/kcp-dev/kcp/pkg/reconciler/tenancy/initialization/apis \
+  "initialization:v1alpha1" \
+  --go-header-file "${SCRIPT_ROOT}"/hack/boilerplate/boilerplate.generatego.txt \
+  --output-base "${SCRIPT_ROOT}" \
+  --trim-path-prefix github.com/kcp-dev/kcp &
+
+for pid in $( jobs -p ); do
```
Comment (stevekuznetsov) [inline hack/update-codegen-clients.sh:55]: This is unrelated but sped up codegen for me substantially. Boy is this thing slow.

File: hack/update-codegen.sh [hunk 4]
```diff
@@ -47,5 +47,12 @@ ${CONTROLLER_GEN} \
     crd \
     rbac:roleName=manager-role \
     webhook \
-    paths="./test/e2e/reconciler/cluster/..." \
-    output:crd:artifacts:config=test/e2e/reconciler/cluster/
+    paths="./test/e2e/fixtures/wildwest/..." \
+    output:crd:artifacts:config=test/e2e/fixtures/wildwest/
```
Comment (stevekuznetsov) [inline hack/update-codegen.sh:51]: Unrelated but this seems to have been broken for a long time, just nobody changed the test type.

File: pkg/admission/reservedcrdgroups/admission.go [hunk 5]
```diff
@@ -70,7 +70,7 @@ func (o *reservedCRDGroups) Validate(ctx context.Context, a admission.Attributes
 		return nil
 	}
 
-	if strings.HasSuffix(crd.Spec.Group, "kcp.dev") {
+	if strings.HasSuffix(crd.Spec.Group, "kcp.dev") && crd.Spec.Group != "initialization.tenancy.kcp.dev" {
```
Comment (stevekuznetsov) [inline pkg/admission/reservedcrdgroups/admission.go:73]: We need something durable here, this works for now ...

File: pkg/reconciler/tenancy/initialization/apibinder/bootstrap.go (hunk 2 above)
Comment (stevekuznetsov) [reply pkg/reconciler/tenancy/initialization/apibinder/bootstrap.go:101]: This one is not in a controller.

**Comment by stevekuznetsov** [comment]
@sttts this is ready for review ... we're missing a _lot_ of testing for the controller reconcile loops, edge cases, negatives, etc ... however for this to be useful by 0.6 release I need to get inheritance into the type system first ... perhaps we can delay the tests until we have that?

File: pkg/reconciler/tenancy/initialization/apibinder/apibinder_initializer_controller_reconcile.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+	var errors []error
+	client := b.kcpClusterClient.Cluster(logicalcluster.From(clusterWorkspace).Join(clusterWorkspace.Name)).ApisV1alpha1().APIBindings()
+	for _, rawBinding := range configuration.Spec.Bindings {
+		binding, err := client.Create(ctx, &rawBinding, metav1.CreateOptions{})
+		if err != nil && !k8serrors.IsAlreadyExists(err) {
+			errors = append(errors, fmt.Errorf("could not create APIBinding %q: %w", rawBinding.Name, err))
+			continue
+		}
+
+		if err := wait.ExponentialBackoffWithContext(ctx, wait.Backoff{
```
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply pkg/reconciler/tenancy/initialization/apibinder/apibinder_initializer_controller_reconcile.go:75]: Added as a TODO

File: pkg/reconciler/tenancy/initialization/apis/initialization/v1alpha1/types.go (hunk 1 above)
Comment (stevekuznetsov) [reply pkg/reconciler/tenancy/initialization/apis/initialization/v1alpha1/types.go:37]: `InitialAPIBinding`...?

File: hack/update-codegen-clients.sh (hunk 3 above)
Comment (sttts) [reply hack/update-codegen-clients.sh:55]: It only reads 10 million lines of go code and parses it on every generator invocation. What do you expect?

File: pkg/admission/reservedcrdgroups/admission.go (hunk 5 above)
Comment (sttts) [reply pkg/admission/reservedcrdgroups/admission.go:73]: /hold

This is wrong.

File: pkg/reconciler/tenancy/initialization/apibinder/apiset_controller.go [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+
+	clusterName := logicalcluster.From(old)
+	name := old.Name
+
+	oldForPatch := initializationv1alpha1.APISet{
+		Status: old.Status,
+	}
+	// to ensure they appear in the patch as preconditions
+	oldForPatch.UID = ""
+	oldForPatch.ResourceVersion = ""
```
Comment (ncdc) [inline pkg/reconciler/tenancy/initialization/apibinder/apiset_controller.go:224]: Can remove

File: pkg/reconciler/tenancy/initialization/apibinder/apiset_controller.go [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+
+const (
+	apiSetControllerName = "kcp-apiset-controller"
+	byExport             = apiSetControllerName + "-byExport"
+)
+
+type CreateAPIDefinitionFunc func(apiResourceSchema *apisv1alpha1.APIResourceSchema, version string, identityHash string) (apidefinition.APIDefinition, error)
+
+// NewAPISetValidator returns a new controller which watches APISets ensures they are valid.
+func NewAPISetValidator(
```
Comment (ncdc) [inline pkg/reconciler/tenancy/initialization/apibinder/apiset_controller.go:57]: (no action required) mismatch between file name and struct - apiset_controller.go vs APISetValidator

File: pkg/reconciler/tenancy/initialization/apibinder/bootstrap.go [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+	gr := metav1.GroupResource{
+		Group:    initialization.GroupName,
+		Resource: "apisets",
+	}
+	crd, err := initializationbootstrap.CustomResourceDefinition(gr)
+	if err != nil {
+		return fmt.Errorf("could not determine CustomResourceDefinition: %w", err)
+	}
+
+	resourceSchema, err := apisv1alpha1.CRDToAPIResourceSchema(crd, invalidCharacters.ReplaceAllString(version.Get().String(), "-")) // TODO: find a sensible thing to use here
```
Comment (ncdc) [inline pkg/reconciler/tenancy/initialization/apibinder/bootstrap.go:69]: Probably need to avoid using `version` for the prefix, as that will change with every commit, won't it? Maybe hard code it to something like v1 or rev1 or ...

File: pkg/reconciler/tenancy/initialization/apibinder/bootstrap.go [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+			klog.Infof("Error waiting for APIExport %q|%q virtual workspace URLs to be ready: %v", installInto.String(), export.Name, err)
+			return false, err
+		}
+		return checkConditionTrueWhileLogging(export, apisv1alpha1.APIExportVirtualWorkspaceURLsReady, fmt.Sprintf("Not done waiting for APIExport %q|%q virtual workspace URLs to be ready", installInto.String(), export.Name)), nil
+	}); err != nil {
+		return fmt.Errorf("could not wait for virtual workspace URLs to be ready on APIExport: %w", err)
+	}
+	klog.Infof("APIExport %q|%q virtual workspace URLs are ready", installInto.String(), export.Name)
+
+	binding, err := apisClient.APIBindings().Create(ctx, &apisv1alpha1.APIBinding{
```
Comment (ncdc) [inline pkg/reconciler/tenancy/initialization/apibinder/bootstrap.go:117]: // TODO remove creating this APIBinding when #1183 is fixed?

File: pkg/reconciler/tenancy/initialization/apibinder/bootstrap.go [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+
+	initializerControllerCfg := rest.AddUserAgent(rest.CopyConfig(cfg), "kcp-apibinder-initializer-bootstrapper-controller")
+
+	initializingWorkspacesCfg := rest.CopyConfig(initializerControllerCfg)
+	initializingWorkspacesCfg.Host = cwt.Status.VirtualWorkspaces[0].URL
+	initializingWorkspacesKcpClusterClient, err := kcpclient.NewClusterForConfig(initializingWorkspacesCfg)
+	if err != nil {
+		return err
+	}
+	initializingWorkspacesKcpInformers := kcpinformer.NewSharedInformerFactory(initializingWorkspacesKcpClusterClient.Cluster(logicalcluster.Wildcard), 10*time.Minute)
```
Comment (ncdc) [inline pkg/reconciler/tenancy/initialization/apibinder/bootstrap.go:187]: Why a 10-minute resync period?

File: pkg/reconciler/tenancy/initialization/apibinder/bootstrap.go [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+		initializationutil.InitializerForType(cwt),
+		initializingWorkspacesKcpClusterClient,
+		initializingWorkspacesKcpInformers.Tenancy().V1alpha1().ClusterWorkspaces(),
+		apiExportInitializationInformers.Initialization().V1alpha1().APISets(),
+	)
+	if err != nil {
+		return err
+	}
+
+	kcpInformers := kcpinformer.NewSharedInformerFactory(kcpClusterClient.Cluster(logicalcluster.Wildcard), 10*time.Minute)
```
Comment (ncdc) [inline pkg/reconciler/tenancy/initialization/apibinder/bootstrap.go:207]: Does this need to be different from the shared factory we already have in the server?

File: pkg/reconciler/tenancy/initialization/apibinder/clusterworkspacetype.yaml [hunk 13]
```diff
@@ -0,0 +1,11 @@
+apiVersion: tenancy.kcp.dev/v1alpha1
+kind: ClusterWorkspaceType
+metadata:
+  name: apibinder
+spec:
+  initializer: true
+  defaultChildWorkspaceType: Universal
+  allowedChildWorkspaceTypes:
+    - Any
```
Comment (ncdc) [inline pkg/reconciler/tenancy/initialization/apibinder/clusterworkspacetype.yaml:9]: (no action required) maybe `*` instead of `Any`?

File: pkg/reconciler/tenancy/initialization/apis/initialization/v1alpha1/doc.go [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// +k8s:deepcopy-gen=package,register
+// +groupName=initialization.tenancy.kcp.dev
```
Comment (ncdc) [inline pkg/reconciler/tenancy/initialization/apis/initialization/v1alpha1/doc.go:18]: Also add `// +k8s:openapi-gen=true`?

File: pkg/reconciler/tenancy/initialization/apibinder/clusterworkspacetype.yaml (hunk 13 above)
Comment (stevekuznetsov) [reply pkg/reconciler/tenancy/initialization/apibinder/clusterworkspacetype.yaml:9]: The use of `Any` predates this PR. `*` does not match our cluster name validating regexes, and it's nice to use the same type here to get the validation ... /shrug

File: pkg/reconciler/tenancy/initialization/apibinder/bootstrap.go (hunk 11 above)
Comment (ncdc) [reply pkg/reconciler/tenancy/initialization/apibinder/bootstrap.go:187]: 0 (no resync), or 10+ hours (see e.g. https://github.com/kcp-dev/kcp/blob/65e631123bc61f1a273422ec9f8137bcbcecd1ad/pkg/server/server.go#L64)

File: pkg/admission/reservedcrdgroups/admission.go (hunk 5 above)
Comment (stevekuznetsov) [reply pkg/admission/reservedcrdgroups/admission.go:73]: Requires #1386

File: pkg/reconciler/tenancy/initialization/apibinder/apibinder_initializer_controller.go [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+		klog.Errorf("failed to determine ClusterWorkspaceType name from reference: %s: %v", clusterWorkspace.Spec.Type, err)
+		return nil // nothing we can do here
+	}
+
+	configuration, err := b.apiSetLister.Get(clusters.ToClusterAwareKey(cwtCluster, cwtName))
+	if err != nil && !apierrors.IsNotFound(err) {
+		klog.Errorf("failed to get APISet %s|%s from lister: %v", cwtCluster, cwtName, err)
+		return nil // nothing we can do here
+	}
+
```
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply pkg/reconciler/tenancy/initialization/apibinder/apibinder_initializer_controller.go:181]: L162 handles it - error that's not NotFound gets errored, NotFound causes us to ignore the key

File: pkg/reconciler/tenancy/initialization/apibinder/apibinder_initializer_controller_reconcile.go [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+			return conditions.IsTrue(binding, apisv1alpha1.InitialBindingCompleted), nil
+		}); err != nil {
+			errors = append(errors, fmt.Errorf("APIBinding %q was never bound: %w", binding.Name, err))
+		}
+	}
+	if len(errors) > 0 {
+		klog.V(2).Infof("Failed initializing APIBindings in %q", clusterName)
+		conditions.MarkFalse(
+			clusterWorkspace,
+			initializedCondition,
```
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply pkg/reconciler/tenancy/initialization/apibinder/apibinder_initializer_controller_reconcile.go:93]: This is local to this particular initializer and there will be one of these per controller, so that did not seem appropriate?

File: pkg/reconciler/tenancy/initialization/apibinder/bootstrap.go (hunk 9 above)
Comment (stevekuznetsov) [reply pkg/reconciler/tenancy/initialization/apibinder/bootstrap.go:69]: Unclear - what if the CRD changes ... ?

File: pkg/reconciler/tenancy/initialization/apibinder/bootstrap.go (hunk 10 above)
Comment (stevekuznetsov) [reply pkg/reconciler/tenancy/initialization/apibinder/bootstrap.go:117]: Intent was for this to live in the root workspace where e.g. organization then wants to use the initializer to add e.g. clusterworkspaces

File: pkg/reconciler/tenancy/initialization/apibinder/bootstrap.go (hunk 12 above)
Comment (stevekuznetsov) [reply pkg/reconciler/tenancy/initialization/apibinder/bootstrap.go:207]: No, I just wanted it to be standalone / simple as possible for the first pass.

File: pkg/reconciler/tenancy/initialization/apibinder/apibinder_initializer_controller_reconcile.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (ncdc) [reply pkg/reconciler/tenancy/initialization/apibinder/apibinder_initializer_controller_reconcile.go:95]: oh, gotcha, thanks

File: pkg/reconciler/tenancy/initialization/apibinder/apibinder_initializer_controller.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (ncdc) [reply pkg/reconciler/tenancy/initialization/apibinder/apibinder_initializer_controller.go:182]: But this is for the APISet. L162 is for the ClusterWorkspace ... ?

[the remark this answers is not in the document]
Comment (ncdc) [reply pkg/reconciler/tenancy/initialization/apibinder/apibinder_initializer_controller.go:182]: (This comment was originally about `configuration`)

**Comment by ncdc** [comment]
Latest changes LGTM. Would still like either a TODO/follow-up issue on kcpInformers or to plumb it through before we merge - sound ok?

File: pkg/reconciler/tenancy/initialization/apibinder/bootstrap.go [hunk 17]
```diff
@@ -0,0 +1,267 @@
...
+		initializationutil.InitializerForType(cwt),
+		initializingWorkspacesKcpClusterClient,
+		initializingWorkspacesKcpInformers.Tenancy().V1alpha1().ClusterWorkspaces(),
+		apiExportInitializationInformers.Initialization().V1alpha1().APISets(),
+	)
+	if err != nil {
+		return err
+	}
+
+	kcpInformers := kcpinformer.NewSharedInformerFactory(kcpClusterClient.Cluster(logicalcluster.Wildcard), 10*time.Hour)
```
Comment (sttts) [inline pkg/reconciler/tenancy/initialization/apibinder/bootstrap.go:198]: 🔴 this is a nogo for in-process. Pass informers as flags.

**Comment by sttts** [comment]
/hold

While the overall value of this API is undisputed, and large parts of the implementation are uncontroversial, the overall factoring, the security constraints, the urgency, the role in our outside of kcp core are not. We better take the necessary time to learn what we want here.

File: pkg/reconciler/tenancy/initialization/apis/initialization/v1alpha1/types.go (hunk 1 above)
Comment (sttts) [reply pkg/reconciler/tenancy/initialization/apis/initialization/v1alpha1/types.go:37]: ClusterWorkspaceTypeAPIBindings

File: pkg/reconciler/tenancy/initialization/apis/initialization/v1alpha1/types.go [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	conditionsv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/apis/conditions/v1alpha1"
+)
+
+// APISet records a set of API Bindings that need to be created
+// and bound before initialization is complete. This configuration
+// applies to all workspaces of the type that shares a name with this
+// object.
```
Comment (sttts) [inline pkg/reconciler/tenancy/initialization/apis/initialization/v1alpha1/types.go:29]: this needs semantics for CWT extension.

File: pkg/reconciler/tenancy/initialization/apis/initialization/v1alpha1/types.go (hunk 1 above)
Comment (davidfestal) [reply pkg/reconciler/tenancy/initialization/apis/initialization/v1alpha1/types.go:37]: Not a fan of types that end-up with an `s`.
What about `ClusterWorkspaceTypeAPIBindingSet` ?

**Comment by stevekuznetsov** [comment]
Superseded by #1793 
/close
